### PR TITLE
Doubly Periodic SCREAM

### PR DESCRIPTION
**Pull request by bogensch**
This PR is an extension of PR #210, which provided much of the infrastructure needed to propagate a single location to the entire domain.  This PR completes development and adds the ability to run SCREAM in doubly-periodic (DP) cloud resolving model (CRM) mode using a planar version of the dynamical core.  This PR provides a working version of DP-SCREAM (with all associated options) but should NOT be used by anyone at this point until a more rigorous scientific validation is completed.  A future PR (if needed) will update to a validated DP-SCREAM.  Run scripts will be provided when DP-SCREAM is ready for general use.

NOTE that a companion CIME PR has also been submitted, which is needed in conjunction with this PR for functioning DP-SCREAM.

Specifically, this PR:
1. renames the ambiguous “iop_mode” flag with “scm_multcols”, which is responsible for providing functionality of the single-column model but to a domain of multiple columns that are identically forced.
2. Adds the configure “dpcrm_mode” and analogous logical “dp_crm”, which is responsible for setting up a doubly periodic CRM.
3. Fixes an issue in ice_grid init, which was not scattering properly the correct surface information for the column of interest to all columns.
4. Introduces namelist variables related to the planar dynamical core that allow the user to set up domain size (se_lx and se_ly) and number of elements (se_ne_x and se_ne_y) on the fly.
5. Renames the logicals that were used to nudge the SCM to observations be more general and introduces the option to nudge T&Q separate from U&V.
6. Implements new nudging routine for the doubly period CRM that nudges off the domain mean.
7. Updates SE initialization to ensure that all columns are initialized identically.
8. Moves the block of code in the SE initialization that introduces random perturbations to AFTER the IOP information is read in to ensure that random perturbations in the initial conditions are not overwritten when DP-SCREAM initializes from the IOP information.
9. Saves the dynamical core grid size information to pass to SHOC.
10. Restores the functionality that was added in PR #210 to initialize land surface identically for all columns when in SCM domain mode.  This was not retained in the renaming from CLM to ELM and was overwritten in the upstream merge.
11. Adds a nightly smoke test for doubly-periodic SCREAM.  Test uses ARM97 case in a small domain for 10 time steps.
12. Add support for exact restarts for DP-SCREAM and adds an exact restart test (ERS).

### CODE REVIEW

**Comment by jgfouca** [comment]
@bogensch , I'm a bit worried these are going to conflict heavily with upstream E3SM. I know we've talked about this in past SCREAM meetings but I can't remember if we came to any agreement on an approach.

**Comment by jgfouca** [comment]
@bogensch , what's the relationship between this PR and your CIME PR https://github.com/ESMCI/cime/pull/3883? Does this PR depend on that CIME PR?

**Comment by bogensch** [comment]
@jgfouca, yes this PR depends on the CIME PR for a properly functioning DP-SCREAM.  The CIME PR is responsible for setting all the necessary flags and dimensions of the domain so that component model domains are compatible with the ATM.

In regards to your concerns about upstream merge.  I also remember talking about this but not sure if we agreed on any approach.  I would expect there to be conflicts, as there was in the last upstream merge in regards to PR #210 (which is related to this current one).  I thought the process of you delegating the files to me that were conflicted and me resolving them went well and efficiently (feel free to disagree).  I don't see there be a substantially greater amount of conflicts than PR #210 introduced, unless there is a particular section of code that's troubling you?

File: components/eam/src/dynamics/se/se_single_column_mod.F90
Comment (celdredsandia) [inline components/eam/src/dynamics/se/se_single_column_mod.F90:116]: Eventually I will add support for non-uniform planar grids, but this should be fine for now.

File: components/homme/src/share/namelist_mod.F90
Comment (celdredsandia) [inline components/homme/src/share/namelist_mod.F90:222]: All these modifications look fine to me! Do we want to add the ability to set Lx/Ly in standalone planar HOMME as well? Since standalone HOMME is only (to my knowledge) ever run with test cases as initial conditions (which have a given Lx/Ly) this is probably not needed...

**Review by celdredsandia — CHANGES_REQUESTED** [review]
The changes look good to me. However, I did not see any tests added for the new functionality. For EAM/E3SM, do these get set up separately in the autotester?

**Comment by jgfouca** [comment]
@bogensch , I'm fine with repeating the process we did in the last upstream merge if you are.

**Comment by bogensch** [comment]
Opening this back up for review/re-review.

**Review by celdredsandia — APPROVED** [review]
Looks good to me now!

File: components/homme/src/share/namelist_mod.F90
Comment (bogensch) [reply components/homme/src/share/namelist_mod.F90:222]: I'll leave this up to you.  I never run HOMME in standalone mode so I think you're better equipped to answer this; but if what you said is in fact true then yes, it is probably not needed.

Comment (celdredsandia) [reply components/homme/src/share/namelist_mod.F90:222]: Let's not worry about it for now- it is easy enough to add the capability if needed.

**Comment by oksanaguba** [comment]
just noticed I was tagged for a review, sorry for the delay, will do soon.

**Comment by oksanaguba** [comment]
@bogensch why would one want functionality "functionality of the single-column model but to a domain of multiple columns that are identically forced"?

**Comment by oksanaguba** [comment]
I am confused whether this PR is about DP+SCREAM or DP+SCM...

**Comment by bogensch** [comment]
DP-SCREAM uses the same intensive observation period (IOP) forcing files that are used for the SCM to force the simulation.  Thus, we take a single point of the globe (at the location of the IOP forcing file being used), and extract that point to the entire domain.  You can essentially think of DP-SCREAM as a domain of SCMs, but with dynamics turned on.  Therefore, a prerequisite to get DP-SCREAM to work was to ensure that the SCM functionality could operate over multiple columns, as the SCM infrastructure was originally designed with the intention that it would only be used over one singular column.  

Most of that functionality (what you refer to as DP+SCM) was accomplished in PR #210 (though this PR does rename a bunch of poorly named flags that were declared in that PR… and restores some functionality in ELM that was lost when the name change from CLM to ELM occurred that was not brought over in our upstream merge).  This PR mostly ties everything together (DP+SCM infrastructure, Chris eldred’s planar dy-core, etc.) and adds some new enhancements/functionality to produce a realized/functioning DP-SCREAM.  A comprehensive and itemized list of what was added can be found on the PR description.

File: components/eam/cime_config/config_component.xml
Comment (crterai) [inline components/eam/cime_config/config_component.xml:132]: Seeing the errors reported on slack caused by the recent upstream merge, `-phys cam5` might need to be replaced with `-phys default`.

File: components/eam/cime_config/config_component.xml
Comment (crterai) [inline components/eam/cime_config/config_component.xml:146]: See comment above about `-phys default`

File: components/elm/src/main/elm_varctl.F90
Comment (crterai) [inline components/elm/src/main/elm_varctl.F90:453]: `elm_varctl_set` is also called in `lnd_comp_esmf`. It doesn't seem that these newly added parameters are specified when it's called there. Is it unneeded in that case?

**Review by crterai — APPROVED** [review]
I left a couple comments/questions for you to review, but apart from the comment about the -phys option, which might be the source of the conflict that's showing up right now, looks good to me!

File: components/eam/src/dynamics/se/stepon.F90
Comment (oksanaguba) [inline components/eam/src/dynamics/se/stepon.F90:542]: Is this unwrapping to a 1d array needed?
Just looking at DYN_PS output in stepon, that does not need 1D array:
`      call outfld('DYN_PS'    ,p_i(:,:,nlevp),npsq,ie)`

**Comment by oksanaguba** [comment]
Just some questions/comments:
```
6.Implements new nudging routine for the doubly period CRM that nudges off the domain mean.
7.Updates SE initialization to ensure that all columns are initialized identically.
8.Moves the block of code in the SE initialization that introduces random perturbations to AFTER the IOP information is read in to ensure that random perturbations in the initial conditions are not overwritten when DP-SCREAM initializes from the IOP information.
9.Saves the dynamical core grid size information to pass to SHOC.
10.Restores the functionality that was added in PR #210 to initialize land surface identically for all columns when in SCM domain mode. This was not retained in the renaming from CLM to ELM and was overwritten in the upstream merge.
```
Is #6 tested?
Is #8 tested? Probably not, since it is not a bfb simulation. If so, there is no way to verify the changes.
Was #10 tested? Probably, not, since it got broken with merge.

If any of this functionality needed, it probably needs a test (except for the perturbation).

File: components/homme/src/share/namelist_mod.F90
Comment (oksanaguba) [inline components/homme/src/share/namelist_mod.F90:997]: Should it be something like abs(Lx) <= tolerance? 
Overall, inputs for planar in homme need some cleaning, this is out of scope for this PR.

**Review by celdredsandia — APPROVED** [review]

**Comment by bogensch** [comment]
Now that the required CIME changes have been merged in, this PR is now finally ready to go.  

@oksanaguba , I think we just need your approval if you have no further concerns.  Though, I realized I never responded to your question from May 3rd; Yes, 6 and 10 have been tested and they are exercised in the nightly smoke and restart tests added.

**Review by oksanaguba — APPROVED** [review]

**Comment by oksanaguba** [comment]
approved, thanks

**Pull request merged**
